File: module_test_1/src/main/org/grynko/nazar/java/Bus.java
package main.org.grynko.nazar.java;

import lombok.SneakyThrows;

import java.util.List;
import java.util.Random;

import static java.lang.Thread.interrupted;

public class Bus implements Runnable {

    private int currentIndex;
    private final List<BusStop> stops;
    private Direction direction;

    public Bus(List<BusStop> stops) {
        this.currentIndex = 0;
        this.stops = stops;
        this.direction = Direction.FORWARD;
    }

    @Override
    public void run() {
        BusStop busStop = stops.get(currentIndex);

        while(!interrupted()) {
            busStop.stop();

            randomSleep();
            busStop = getNextBusStop();
        }
    }

    private BusStop getNextBusStop() {
        currentIndex = nextIndex();
        return stops.get(currentIndex);
    }

    private int nextIndex() {
        int stopAmount = stops.size();
        int nextIndex = currentIndex;

        if(currentIndex == stopAmount - 1) {
            nextIndex--;
            direction = Direction.BACKWARD;
        }
        else if(currentIndex == 0) {
            nextIndex++;
            direction = Direction.FORWARD;
        }
        else {
            switch (direction) {
                case FORWARD: {
                    nextIndex++;
                    break;
                }
                case BACKWARD: {
                    nextIndex--;
                    break;
                }
            }
        }

        return nextIndex;
    }

    @SneakyThrows
    private void randomSleep() {
        Random random = new Random();
        Thread.sleep(random.nextInt(200) + 300);
    }

}

File: module_test_1/src/main/org/grynko/nazar/go/bus/bus.go
package bus

import (
	"Lab_3/module_test/bus_stop"
	"math/rand"
	"time"
)

const (
	BACKWARD = iota
	FORWARD
)

type Bus struct {
	stops     []bus_stop.BusStop
	currIndex int
	direct    int
	name      string
}

func Init(stops []bus_stop.BusStop, name string) *Bus {
	bus := new(Bus)

	bus.stops = stops
	bus.currIndex = 0
	bus.direct = FORWARD
	bus.name = name

	return bus
}

func (bus *Bus) Run() {
	busStop := bus.stops[bus.currIndex]

	for true {
		busStop.Stop(bus.name)

		time.Sleep(time.Duration(rand.Intn(200)+300) * time.Millisecond)
		busStop = bus.getNextBusStop()
	}
}

func (bus *Bus) nextIndex() int {
	stopAmount := len(bus.stops)
	nextIndex := bus.currIndex

	if bus.currIndex == stopAmount-1 {
		nextIndex -= 1
		bus.direct = BACKWARD
	} else if bus.currIndex == 0 {
		nextIndex++
		bus.direct = FORWARD
	} else {
		switch bus.direct {
		case FORWARD:
			{
				nextIndex++
				break
			}
		case BACKWARD:
			{
				nextIndex--
				break
			}
		}
	}

	return nextIndex
}

func (bus *Bus) getNextBusStop() bus_stop.BusStop {
	bus.currIndex = bus.nextIndex()
	return bus.stops[bus.currIndex]
}
